lib/common/ipc_client.c

```diff
@@ -31,6 +31,672 @@
+ * \param[out] dispatch_type  How IPC replies should be dispatched
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+pcmk_connect_ipc(pcmk_ipc_api_t *api, enum pcmk_ipc_dispatch dispatch_type)
+{
+    int rc = pcmk_rc_ok;
+
+    if ((api == NULL) || (api->ipc == NULL)) {
+        crm_err("Cannot connect to uninitialized API object");
+        return EINVAL;
+    }
+
+    if (crm_ipc_connected(api->ipc)) {
+        crm_trace("Already connected to %s IPC API", pcmk_ipc_name(api, true));
+        return pcmk_rc_ok;
+    }
+
+    api->dispatch_type = dispatch_type;
+    switch (dispatch_type) {
+        case pcmk_ipc_dispatch_main:
+            rc = connect_with_main_loop(api);
+            break;
+
+        case pcmk_ipc_dispatch_sync:
+        case pcmk_ipc_dispatch_poll:
+            rc = connect_without_main_loop(api);
+            break;
+    }
+    if (rc != pcmk_rc_ok) {
+        return rc;
+    }
+
+    if ((api->cmds != NULL) && (api->cmds->post_connect != NULL)) {
+        rc = api->cmds->post_connect(api);
+        if (rc != pcmk_rc_ok) {
+            crm_ipc_close(api->ipc);
+        }
+    }
+    return rc;
+}
+
+/*!
+ * \brief Disconnect an IPC API instance
+ *
+ * \param[in]  api  IPC API connection
+ *
+ * \return Standard Pacemaker return code
+ *
+ * \note If the connection is attached to a main loop, this function should be
+ *       called before quitting the main loop, to ensure that all memory is
+ *       freed.
+ */
+void
+pcmk_disconnect_ipc(pcmk_ipc_api_t *api)
+{
+    if ((api == NULL) || (api->ipc == NULL)) {
+        return;
+    }
+    switch (api->dispatch_type) {
+        case pcmk_ipc_dispatch_main:
+            {
+                mainloop_io_t *mainloop_io = api->mainloop_io;
+
+                // Make sure no code with access to api can use these again
+                api->mainloop_io = NULL;
+                api->ipc = NULL;
+
+                mainloop_del_ipc_client(mainloop_io);
+                // After this point api might have already been freed
+            }
+            break;
+
+        case pcmk_ipc_dispatch_poll:
+        case pcmk_ipc_dispatch_sync:
+            {
+                crm_ipc_t *ipc = api->ipc;
+
+                // Make sure no code with access to api can use ipc again
+                api->ipc = NULL;
+
+                // This should always be the case already, but to be safe
+                api->free_on_disconnect = false;
+
+                crm_ipc_destroy(ipc);
+                ipc_post_disconnect(api);
+            }
+            break;
+    }
+}
+
+/*!
+ * \brief Register a callback for IPC API events
+ *
+ * \param[in] api          IPC API connection
+ * \param[in] callback     Callback to register
+ * \param[in] userdata     Caller data to pass to callback
+ */
+void
+pcmk_register_ipc_callback(pcmk_ipc_api_t *api, pcmk_ipc_callback_t cb,
+                           void *user_data)
+{
+    if (api != NULL) {
+        api->cb = cb;
```

Comment:
If api->cb is already set, would we want that to be an error?  If not, perhaps it warrants a mention in the doc string.

Reply:
I think it's reasonable to call it multiple times to change the callback as desired

Reply:
I thought it might be.  I also thought it might be worth a mention in the docs just so we remember it's a thing you can do.